File: src/Alimer/Engine/Component.cs
// Copyright (c) Amer Koleci and Contributors.
// Licensed under the MIT License (MIT). See LICENSE in the repository root for more information.

using Alimer.Serialization;

namespace Alimer.Engine;

/// <summary>
/// Defines a component that can be attached to an <see cref="Alimer.Engine.Entity"/>.
/// </summary>
[Meta]
public abstract partial class Component : ISerializable
{
    /// <summary>
	/// The version of the component.
	/// </summary>
    public virtual int ComponentVersion => 0;

    public Entity? Entity { get; internal set; }

    public virtual bool IsEnabled { get; set; } = true;

    public void Serialize(Serializer serializer)
    {
        string componentTypeName = GetType().Name;
        //using ObjectSerializer objectSerializer = serializer.BeginObject();
        //serializer.WriteType(componentTypeName);
        //serializer.WriteVersion(ComponentVersion);
        //serializer.Write(Keys.IsEnabled, IsEnabled);
        //OnSerialize(serializer);
    }

    public void Deserialize(ObjectDeserializer deserializer)
    {
        //IsEnabled = deserializer.ReadBool(Keys.IsEnabled, IsEnabled);
    }

    protected virtual void OnSerialize(ObjectSerializer serializer)
    {

    }

    internal static class Keys
    {
        public const string Id = "Id";
        public const string IsEnabled = nameof(Component.IsEnabled);
    }
}

File: Source/Alimer/Scene/Component.cpp
// Copyright (c) Amer Koleci and Contributors.
// Licensed under the MIT License (MIT). See LICENSE in the repository root for more information.


#include "Alimer/Scene/Component.h"
#include "Alimer/Scene/Entity.h"
#include "Alimer/Scene/Scene.h"

using namespace Alimer;

Entity* Component::GetEntity() const
{
    return _entity;
}

[[nodiscard]] UUID Component::GetID() const
{
    return _id;
}

[[nodiscard]] Scene* Component::GetScene() const
{
    return _entity ? _entity->GetScene() : nullptr;
}

void Component::SetEntity(Entity* value)
{
    if (_entity == value)
        return;

    _entity = value;
}

void Component::SetID(UUID value)
{
    _id = value;
}

void Component::SetEnabled(bool value)
{
    if (_enabled == value)
        return;

    _enabled = value;

    Scene* scene = GetScene();
    if (scene != nullptr)
    {
        //scene->OnComponentEnabledChanged(this);
    }
}
